### PR TITLE
Added support to modify service startup

### PR DESCRIPTION
**Pull request by Cicco0**
I needed the possibility to modify the startup service, so I put this in. Now you are able to start bind in IPv4/IPv6 mode or passing some other variables.

In addition I had to fix some issues with concat_basedir to run rpsec tests locally.

### CODE REVIEW

File: manifests/server/startup.pp [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# == Define: dns::server::startup
```
Comment (solarkennedy) [inline manifests/server/startup.pp]: I think we are ready to need this. I don't really like the name "startup" though, it makes me think it handles init script.

How about class: dns::server::defaults_file ?

File: manifests/server/startup.pp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#   Set string "yes", "no" or leave empty. Default: empty
+#
+# === Examples
+#
+#  dns::server::startup { '/etc/default/bind9':
+#    resolvconf => 'no',
+#    options    => '-u bind'
+#  }
+#
+class dns::server::startup (
```
Comment (solarkennedy) [inline manifests/server/startup.pp]: Is it a define or a class? Should be a class, we only have one o them.

File: manifests/server/startup.pp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+# [*keytab_file*]
+#   Must be an absolute path to set up the keytab file
+#
+# [*disable_zone_checking*]
+#   If you set this option to 'yes' then initscript doesn't perform named-checkzone
+#   Set string "yes", "no" or leave empty. Default: empty
+#
+# === Examples
+#
+#  dns::server::startup { '/etc/default/bind9':
```
Comment (solarkennedy) [inline manifests/server/startup.pp]: I don't want to make the same mistake as the options file. This shouldn't be a defined type that users have to "remember" to add to their manifest. init.pp should include this.

Comment (Cicco0) [inline manifests/server/startup.pp]: Okay, i will rewrite the code from a class to a define type

File: manifests/server/startup.pp (hunk 2 above)
Comment (Cicco0) [inline manifests/server/startup.pp]: It is a class, the documentation is wrong.

File: manifests/server/startup.pp (hunk 1 above)
Comment (Cicco0) [inline manifests/server/startup.pp]: How about class dns::server::defaults? The definition of **defaults_file** will be used in the code to set the filename.

Comment (solarkennedy) [inline manifests/server/startup.pp]: That sounds fine

**Comment by Cicco0** [comment]
Now it is renamed and corrected. Do you repair the merge conflict?

**Comment by solarkennedy** [comment]
It is usually up to the pull requested to fix merge conflicts and bring the PR into a mergable form.

**Comment by Cicco0** [comment]
The merge conflicts are solved :-)

File: manifests/init.pp [hunk 4]
```diff
@@ -6,4 +6,5 @@
   # include dns::install
   # include dns::config
   # include dns::service
+  include dns::server::default
```
Comment (solarkennedy) [inline manifests/init.pp]: Shouldn't this be in dns::server::install? Seems out of place here

**Comment by Cicco0** [comment]
Ok is fixed and was moved to "manifest/server/install.pp".

File: manifests/server/default.pp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  $rootdir               = undef,
+  $enable_zone_write     = undef,
+  $enable_sdb            = undef,
+  $disable_named_dbus    = undef,
+  $keytab_file           = undef,
+  $disable_zone_checking = undef,
+
+) inherits dns::server::params {
+
+  if ! defined(Class['::dns::server']) {
```
Comment (solarkennedy) [inline manifests/server/default.pp]: This is incorrect now, this class should be safe to include no matter what

File: manifests/server/install.pp [hunk 6]
```diff
@@ -1,5 +1,7 @@
 # == Class dns::server
 #
+include dns::server::default
```
Comment (solarkennedy) [inline manifests/server/install.pp]: This is not correct, you can't put an include outside of a class definition

Comment (solarkennedy) [inline manifests/server/install.pp]: The default class should be chained after the package.
The package should be installed first, then the default class will replace the stock defaults file in that order.

Comment (Cicco0) [inline manifests/server/install.pp]: I would now solve it like this in the file **server.pp**:
```
class dns::server {
  class { 'dns::server::install': } ->
  class { 'dns::server::config': } ->
  class { 'dns::server::default': } ~>
  class { 'dns::server::service': }
}
```
Do you agree with this changement?

Comment (jearls) [inline manifests/server/install.pp]: You can do this in the puppet 4 / future parser, does this not work in the traditional parser?

Kyle Anderson wrote:
> In manifests/server/install.pp <https://github.com/ajjahn/puppet-dns/pull/130#discussion_r33468120>:
>
>> @@ -1,5 +1,7 @@
>>  # == Class dns::server
>>  #
>> +include dns::server::default
>
> This is not correct, you can't put an include outside of a class definition
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/ajjahn/puppet-dns/pull/130/files#r33468120>.
>

-- 
Johnson Earls
darkfoxprime@gmail.com

Comment (solarkennedy) [inline manifests/server/install.pp]: It does, but you shouldn't. All resources need to be contained in some sort of class, not just setting out there for the parser to suck up as it reads the file.

Comment (solarkennedy) [inline manifests/server/install.pp]: @Cicco0 yea I think that would work, but I think I actually prefer it to be in the config class, as it is a config file?

Comment (Cicco0) [inline manifests/server/install.pp]: @solarkennedy I've included it:
`include dns::server::default`

Now everything should be fine. If so, you need a rebase?

Comment (solarkennedy) [inline manifests/server/install.pp]: instlal.pp:3 still needs to be removed. It is a include line outside a class definition and will be included by simply parsing this file, which is surprising and impossible to test.

Comment (Cicco0) [inline manifests/server/install.pp]: Was removed and tests were positive.

**Comment by solarkennedy** [comment]
Good.

We have an acceptance test suite using beaker that is now failing. We don't run these on Travis as they are too intensive.

You can run these tests with `bundle exec rake beaker` per the contributing doc:
https://github.com/ajjahn/puppet-dns/blob/master/CONTRIBUTING.md#integration-tests

I ran them here and here are the failures:
http://pastebin.com/JJWqzBqA

**Comment by Cicco0** [comment]
I'll try it. Apparently I do not get it to work "out of the box".

**Comment by Cicco0** [comment]
Now, after several more days, the problem could be solved. All tests were carried out and I get no error.

**Comment by solarkennedy** [comment]
Good. Can you pastebin the output?

**Comment by Cicco0** [comment]
Sure:
 - bundle exec rake test
 - - http://pastebin.com/Hw8JGfVk
- bundle exec rake spec
 - - http://pastebin.com/5SZYS1Zq
. bundle exec rake beaker
 - - http://pastebin.com/LQbYJ0D8

**Comment by solarkennedy** [comment]
Thank you!

**Pull request merged**